### ISSUE TITLE
Apodization Functionality

### ISSUE THREAD

**Issue by smartalecH** [stevengj/meep#375]
It might be nice to be able to pass a user defined apodization function to the `get_flux_data()` and `get_eigenmode_coefficients()` functions. For high q devices (i.e. ring resonators) it's important for the device to "ring out" before the simulation ends. This can be somewhat tricky to simulate with the time domain solver. One pseudo-remedy is to window the time domain data saved at the various ports (flux or eigenmode) before taking the DFT. This helps reduce the sinc-like ringing often observed in the frequency domain.

**Comment by stevengj** [comment]
User-defined window functions change the nature of the windowing distortion, but I'm not sure it fundamentally improves it.  (e.g. instead of getting sinc-like patterns, you "just" get a distorted central peak.   The nice thing about getting the sinc side lobes is that at least it is obvious that you truncated the simulation too early.  With other windowing functions the errors aren't eliminated, only disguised.  Is that better?)

The ideal thing here would probably be to do Pade extrapolation.

**Comment by smartalecH** [comment]
You're absolutely right -- changing the window simply changes the distortion, but doesn't necessarily solve the problem. Depending on your application, however, this may be sufficient. 

Performing a Pade expansion and then extrapolating the tail is a creative idea. I've done similar things with impulse response estimation.

**Comment by stevengj** [comment]
See W.-H. Guo, W.-J. Li, Y.-Z. Huang, Computation of resonant frequencies and quality factors of cavities by FDTD technique and Padé approximation, IEEE Microwave and Wireless Comp. Lett. 11 (5) (2001) 223–225.  S. Cabay, D.-K. Choi, Algebraic computations of scaled Padé fractions, SIAM J. Comput. 15 (1) (1986) 243–270.

**Comment by stevengj** [comment]
There are also extensions of the filter-diagonalization method (used by harminv) to extrapolate spectra.

**Comment by stevengj** [comment]
The computation of Pade approximants is pretty subtle and there are a lot of algorithms.  One starting point might be: https://people.maths.ox.ac.uk/trefethen/publication/PDF/2011_144.pdf
